File: src/core/base/common/uuid.ts
/**
 * Regular expression to validate if a string is a valid UUID (version 4).
 * A UUID is a 128-bit number used to uniquely identify information.
 * The pattern checks the format: xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx.
 */
const UUID_PATTERN =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

/**
 * Checks if the provided value is a valid UUID.
 *
 * @param value - The string to be checked.
 * @returns true if the value matches the UUID pattern, false otherwise.
 */
export function isValidUUID(value: string): boolean {
  return UUID_PATTERN.test(value);
}

/**
 * Declares the `crypto` object with optional methods for random number generation.
 * `crypto.getRandomValues` fills typed arrays with cryptographically strong random values.
 * `crypto.randomUUID` generates a random UUID.
 */
declare const crypto:
  | undefined
  | {
      getRandomValues?(data: Uint8Array): Uint8Array;
      randomUUID?(): string;
    };

/**
 * Generates a UUID (version 4).
 * - If `crypto.randomUUID` is available, it uses it for secure UUID generation.
 * - Otherwise, it falls back to a custom implementation using `crypto.getRandomValues` or manually generated random values.
 *
 * @returns A valid UUID string.
 */
export const generateUUID = (function (): () => string {
  // Use `randomUUID` if available
  if (typeof crypto === "object" && typeof crypto.randomUUID === "function") {
    return crypto.randomUUID.bind(crypto);
  }

  // Fallback: Use `getRandomValues` if available
  let getRandomValues: (array: Uint8Array) => Uint8Array;
  if (
    typeof crypto === "object" &&
    typeof crypto.getRandomValues === "function"
  ) {
    getRandomValues = crypto.getRandomValues.bind(crypto);
  } else {
    // Fallback for environments without `crypto`
    getRandomValues = function (array: Uint8Array): Uint8Array {
      for (let i = 0; i < array.length; i++) {
        array[i] = Math.floor(Math.random() * 256); // Generate random byte
      }
      return array;
    };
  }

  // Prepare hex lookup table for quick conversion from byte to hex string
  const hexLookup: string[] = Array.from({ length: 256 }, (_, i) =>
    i.toString(16).padStart(2, "0")
  );

  // Create a buffer to hold 16 random bytes (128 bits)
  const randomBytes = new Uint8Array(16);

  /**
   * Generates a UUID (version 4) by manually setting version bits and random values.
   *
   * @returns A valid UUID string in the form xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx.
   */
  return function generateUUID(): string {
    // Fill randomBytes with random values
    getRandomValues(randomBytes);

    // Set UUID version and variant bits (version 4)
    randomBytes[6] = (randomBytes[6] & 0x0f) | 0x40; // Set version to 4
    randomBytes[8] = (randomBytes[8] & 0x3f) | 0x80; // Set variant to RFC4122

    // Convert bytes to UUID string format
    return [
      hexLookup[randomBytes[0]],
      hexLookup[randomBytes[1]],
      hexLookup[randomBytes[2]],
      hexLookup[randomBytes[3]],
      "-",
      hexLookup[randomBytes[4]],
      hexLookup[randomBytes[5]],
      "-",
      hexLookup[randomBytes[6]],
      hexLookup[randomBytes[7]],
      "-",
      hexLookup[randomBytes[8]],
      hexLookup[randomBytes[9]],
      "-",
      hexLookup[randomBytes[10]],
      hexLookup[randomBytes[11]],
      hexLookup[randomBytes[12]],
      hexLookup[randomBytes[13]],
      hexLookup[randomBytes[14]],
      hexLookup[randomBytes[15]],
    ].join("");
  };
})();

File: src-tauri/src/core/base/uuid.rs
use rand::Rng;
use regex::Regex;

/// Regular expression pattern to validate if a string is a valid UUID (version 4).
/// A UUID is a 128-bit number used to uniquely identify information.
/// This pattern checks for the format: xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx.
fn is_valid_uuid(value: &str) -> bool {
    let uuid_pattern =
        Regex::new(r"^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$")
            .expect("Failed to compile regex pattern");
    uuid_pattern.is_match(value)
}

/// Generates a UUID (version 4) in the format xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx.
///
/// - This function uses cryptographically strong random values where available.
/// - Sets the version to 4 and the variant to RFC4122 for UUIDs.
///
/// Returns a valid UUID string.
fn generate_uuid() -> String {
    // Create an array to hold 16 random bytes (128 bits)
    let mut random_bytes = [0u8; 16];
    let mut rng = rand::thread_rng();

    // Fill `random_bytes` with random values
    rng.fill(&mut random_bytes);

    // Set version and variant bits
    random_bytes[6] = (random_bytes[6] & 0x0f) | 0x40; // Set version to 4
    random_bytes[8] = (random_bytes[8] & 0x3f) | 0x80; // Set variant to RFC4122

    // Convert bytes to UUID string format
    format!(
        "{:02x}{:02x}{:02x}{:02x}-{:02x}{:02x}-{:02x}{:02x}-{:02x}{:02x}-{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}",
        random_bytes[0], random_bytes[1], random_bytes[2], random_bytes[3],
        random_bytes[4], random_bytes[5],
        random_bytes[6], random_bytes[7],
        random_bytes[8], random_bytes[9],
        random_bytes[10], random_bytes[11], random_bytes[12], random_bytes[13], random_bytes[14], random_bytes[15]
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_uuid_format() {
        let uuid = generate_uuid();
        // Check if the generated UUID is in the correct format
        assert!(
            is_valid_uuid(&uuid),
            "Generated UUID is not in a valid format"
        );
    }

    #[test]
    fn test_is_valid_uuid_valid_uuid() {
        // A sample valid UUID
        let valid_uuid = "123e4567-e89b-12d3-a456-426614174000";
        assert!(is_valid_uuid(valid_uuid), "Expected valid UUID to pass");
    }

    #[test]
    fn test_is_valid_uuid_invalid_uuid_wrong_format() {
        // Invalid UUID (wrong format)
        let invalid_uuid = "12345678-1234-1234-1234-1234567890";
        assert!(
            !is_valid_uuid(invalid_uuid),
            "Expected invalid UUID to fail"
        );
    }

    #[test]
    fn test_is_valid_uuid_invalid_uuid_wrong_version() {
        // Invalid UUID (version other than 4)
        let invalid_uuid = "123e4567-e89b-12d3-b456-426614174000"; // Wrong version bits
        assert!(
            !is_valid_uuid(invalid_uuid),
            "Expected invalid UUID to fail due to wrong version"
        );
    }

    #[test]
    fn test_is_valid_uuid_empty_string() {
        // Empty string should fail validation
        let empty_string = "";
        assert!(
            !is_valid_uuid(empty_string),
            "Expected empty string to be invalid UUID"
        );
    }

    #[test]
    fn test_is_valid_uuid_random_strings() {
        // Random strings that are not UUIDs
        let random_strings = vec![
            "not-a-uuid",
            "12345678",
            "abcdefgh-ijkl-mnop-qrst-uvwxyzabcdef",
            "123e4567e89b12d3a456426614174000", // Missing hyphens
        ];

        for random_string in random_strings {
            assert!(
                !is_valid_uuid(random_string),
                "Expected random string '{}' to be invalid UUID",
                random_string
            );
        }
    }

    #[test]
    fn test_generate_multiple_uuids_unique() {
        // Generate multiple UUIDs and ensure they are unique
        let uuid1 = generate_uuid();
        let uuid2 = generate_uuid();
        assert_ne!(uuid1, uuid2, "Expected generated UUIDs to be unique");
    }
}
